### DISCUSSION TITLE
USE_DEEPSLEEP  and Display not work together?

### DISCUSSION THREAD

**Discussion by Subnum12** [arendst/Tasmota#18253] [Q&A] [answered]
Hello, 

I just noticed strange behavior with USE_DISPLAY_SSD1306 and USE_DEEPSLEEP.  
USE_DEEPSLEEP should be included in [tasmota](https://github.com/arendst/Tasmota/blob/development/BUILDS.md). After adding `#define USE_DISPLAY_SSD1306` to my _user_config_override.h_ DEEPSLEEP disappears from binary (command not found).

manually adding `define USE_DEEPSLEEP ` to user_config_override.h makes DEEPSLEEP available, but now SSD1306 is missing now. Is there any way to have a binary with both options?

tasmota/user_config_override.h:
```
#ifndef USE_AHT2x
#define USE_AHT2x       // [I2cDriver43] Enable AHT20/AM2301B instead of AHT1x humidity and temperature sensor (I2C address 0x38) (+0k8 code)
#endif

#ifndef USE_SCRIPT
#define USE_SCRIPT  // adds about 17k flash size, variable ram size
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif

#ifndef USE_DISPLAY_SSD1306
#define USE_DISPLAY_SSD1306
#endif

#endif  // _USER_CONFIG_OVERRIDE_H_
```

**Comment by sfromis** [comment]
You did not give much info about exactly what you're doing, like what build environment you select when building. Assuming ESP8266.

If you build using the environment for displays or LVGL, `USE_DEEPSLEEP` is undefined, as this option is not seen as useful for displays. And if you build other environments than the one for displays, USE_DISPLAYS_SSD1306 is *not* enabled by default. The general idea is that if you want "mixed" environments, then you should start with the base "tasmota" build environment, and add the feature combo you want.

**Comment by Subnum12** [comment]
I'm using [Master Release](https://gitpod.io/#https://github.com/arendst/Tasmota/tree/master) with no changes except 
```
board                   = esp8266_4M2M
board_build.f_cpu       = 160000000L
board_build.f_flash     = 40000000L 
```
in platformio_override.ini and the above changes to user_config_override.h. Intention is to modify the base "tasmota" build by adding sensor, script and display support. With sensor and script it worked like charm, with display either display or deep sleep get not added. 
I know deep sleep is not common use-case with display, however here it is required.

As from what I read I wonder if just adding `#define USE_DISPLAY_SSD1306` to my user_config_override.h is the right way to add SSD1306 display support to base "tasmota" build as that seems not to work for everyone?

**Reply by barbudor** [reply]
USE_DISPLAY_SSD1306 also requires the addition of the more general USE_DISPLAY
It also requires USE_I2C but that already included in base tasmota

AHT2x seem to be exclusive of AHT1x so you need to undef USE_AHT1x

And USE_DEEPSLEEP is already in the standard tasmota 

```C++
#ifdef USE_AHT1x
#undef USE_AHT1x       // [I2cDriver43] Enable AHT10/15 humidity and temperature sensor (I2C address 0x38, 0x39) (+0k8 code)(+0k8 code)
#endif
#ifndef USE_AHT2x
#define USE_AHT2x       // [I2cDriver43] Enable AHT20/AM2301B instead of AHT1x humidity and temperature sensor (I2C address 0x38) (+0k8 code)
#endif

#ifndef USE_SCRIPT
#define USE_SCRIPT  // adds about 17k flash size, variable ram size
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif

#ifndef USE_DISPLAY
#define USE_DISPLAY
#endif
#ifndef USE_DISPLAY_SSD1306
#define USE_DISPLAY_SSD1306
#endif
```

**Reply by barbudor** [reply]
Looking better, `#undef USE_AHT1x` doesn't seem mandatory but still they are exclusive

**Comment by Subnum12** [comment, marked as answer]
@barbudor yes, you are right. Actually that makes the difference:

```
#ifndef USE_DISPLAY
#define USE_DISPLAY
#endif
```
